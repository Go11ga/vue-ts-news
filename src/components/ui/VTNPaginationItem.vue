<template>
    <router-link
        :to="path"
        class="vtn-pagination__item"
        :class="{'vtn-pagination__item--active' : active}"
        @click="onPageChange"
    >
        {{ label }}
    </router-link>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
    name: 'RSSPaginationItem',

    props: {
        label: {
            type: String,
            required: true
        },

        value: {
            type: Number,
            required: true
        },

        active: {
            type: Boolean,
            default: false
        }
    },

    computed: {
        path () {
            const { name, query, params } = this.$route
            const route = {
                name,
                query,
                params: {
                    ...params,
                    page: this.value
                }
            }
            return route
        }
    },

    methods: {
        onPageChange() {
            window.scrollTo({
                top: 0,
                left: 0,
                behavior: 'smooth'
            })
        }
    }
})
</script>
