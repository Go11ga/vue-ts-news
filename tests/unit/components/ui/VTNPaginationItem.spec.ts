import { shallowMount } from '@vue/test-utils'
import VTNPaginationItem from '@/components/ui/VTNPaginationItem.vue'

describe('VTNPaginationItem', () => {
    let wrapper = shallowMount(VTNPaginationItem, {
        global: {
            stubs: ['router-link'],
            mocks: {
                $route: (route: any) => route
            }
        },
        props: {
            label: 'Label',
            value: 1,
            active: false
        }
    })

    it('dummy name test', () => {
        expect(wrapper.vm.$options.name).toMatch('RSSPaginationItem')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('отрисовка и пропсы', async () => {
        expect(wrapper.findAll('.vtn-pagination__item')[0].classes()).not.toContain('vtn-pagination__item--active')

        wrapper.setProps({
            label: 'Label',
            value: 1,
            active: true
        })

        await wrapper.vm.$nextTick()

        expect(wrapper.findAll('.vtn-pagination__item')[0].classes()).toContain('vtn-pagination__item--active')
    })

    it('computed path', () => {
        const localThis = {
            $route: {
                name: 'name',
                query: 'query',
                params: {
                    param_01: 'param_01',
                    param_02: 'param_02'
                }
            },
            value: 1
        }

        const component: any = VTNPaginationItem

        expect(component.computed.path.call(localThis)).toEqual({
            name: 'name',
            query: 'query',
            params: { param_01: 'param_01', param_02: 'param_02', page: 1 }
        })
    })

    it('клик по ссылке вызывает метод onPageChange', async () => {
        window.scrollTo = jest.fn()

        wrapper.find('.vtn-pagination__item').trigger('click')

        await wrapper.vm.$nextTick()

        expect(window.scrollTo).toHaveBeenCalledWith({
            top: 0,
            left: 0,
            behavior: 'smooth'
        })
    })
})