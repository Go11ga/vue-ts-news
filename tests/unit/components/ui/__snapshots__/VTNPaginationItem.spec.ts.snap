// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTNPaginationItem snapshot 1`] = `
<router-link-stub
  class="vtn-pagination__item"
  to="[object Object]"
/>
`;
